GraphQL-Schema-Design.md

```diff
@@ -0,0 +1,142 @@
+      # Don’t do this:
+      location_city
+      location_country
+      # Instead do this:
+      location {
+        city
+        country
+      }
+    }
+   }
+  ```
+
+* Design the schema around first class domain-models, not functional details.
+
+  For instance, rather than mimicking a backend endpoint that allows one to filter artworks by defining a
+  `filter_artworks` field that has a nested artworks connection, expose the ability to filter artworks in a plain
+  `artworks` connection field instead.
+
+  The key aspect here is that we are trying to expose **artworks**, which are the same as artworks retrieved through
+  other means, only the ‘feed’ they are retrieved from is different and that’s an implementation detail, there is no
+  such model as **filter artworks**.
+
+  ```graphql
+  # Bad
+  {
+    filter_artworks(aggregations: [TOTAL]) {
+      counts {
+        total
+      }
+      artworks(first: 10) {
+        edges {
+          node {
+            title
+          }
+        }
+      }
+    }
+  }
+
+  # Good
+  {
+    artworks(aggregations: [TOTAL], first: 10) {
+      counts {
+        total
+      }
+      edges {
+        node {
+          title
+        }
+      }
+    }
+  }
+  ```
+
+## Root fields
+
+* These are entry points into the graph that is Artsy’s data set.
+
+* Not all types necessarily make sense as root fields.
+  - In general only domain models should be exposed as root fields. An ‘artwork’ is such a model, Artsy’s system time
+    (for clock synchronization) probably is not.
+
+* For the cases where there is no root field, but you still need to be able to retrieve an arbitrary node of the graph
+  (e.g. when you need to refetch a node without needing to refetch all parent nodes along the path from the root to said
+  node), there is the special [`node` root field](https://facebook.github.io/relay/graphql/objectidentification.htm).
+
+  In short, this field is able to retrieve an arbitrary node by using a special ‘global’ ID, which has data encoded
+  needed for our system to know the type of that entity and how to retrieve it. For example, based on just
+  `banksy-champagne-formica-flag` the system wouldn’t be able to know what type of entity this refers to, a global ID
+  would rather encode it like `Artwork:banksy-champagne-formica-flag`.
+
+## Lists
+
+It is undesirable to have multiple fields that semantically refer to the same data. So rather than defining both e.g. an
+`artworks` field _and_ an `artworks_connection`, simply choose one form applicable to the data and call it `artworks`.
+
+* A paginated list (such as most associated types) should use ‘connections’ (see
+  [the spec](https://facebook.github.io/relay/graphql/connections.htm) and
+  [this blog post](https://dev-blog.apollodata.com/explaining-graphql-connections-c48b7c3d6976)).
+  - These are more forward-compatible, as they allow for adding metadata to the association itself and the ‘edge’ (the
```

Comment:
The win is that the schema models the data like a graph. If you have a need to access these types by ID then there’s no reason why that couldn’t be added, but in those cases it would probably not be a (paginated) list.

Reply:
I guess my follow up would be, why is modeling the data like a graph important, in this sense? I'm not disagreeing, just wondering if there's a pragmatic reason this is particularly effective, rather than thinking of the edge as a node with two "belongs-to" links. 

I'm just thinking, in Ohm, I'd want to paginate the `Bidder`s in a `Sale`. But to use the edge concept, I'd paginate the `User`s but consume only the edges. That feels a little awkward to me, but I'm still open to the idea that it's worth it. Just curious how.

Reply:
It’s all about browseability of the schema. I get what you mean that just consuming the egde feels a little awkward in graph theory, but in a pragmatic sense I would totally expect the users that participate in a sale to be in the same part of the graph.

Btw, you don’t need to call the field `users`, I would expect it to look like this:

```graphql
{
  sale {
    bidders {
      edges {
        # List of Bidder type
        node {
          # User type
        }
      }
    }
  }
}
```

I would not expect:

```graphql
{
  sale {
    users {
      # List of User type
    }
    bidders {
      # List of Bidder type
    }
  }
```

Because, what does `sale.users` really mean on its own? They are really just bidders in a sale.

Reply:
I should really stress that I care very little for concepts such as graph theory in a purist way and all of my opinions reflect pragmatic usage. MPs schema that has historically _not_ made use of these concepts where it could has seen a proliferation of fields that have not helped to make the schema intuitive in some cases.

Reply:
Ahhh, now that's an interesting point -- the name of the connection may come from the edge domain object rather than the destination. Hadn't thought of it that way. I'll add that thought to the document. I'm reading [an Apollo post on Relay's connections](https://dev-blog.apollodata.com/understanding-pagination-rest-graphql-and-relay-b10f835549e7), and it's kind of like calling the connection from user's who are friends `friendships`. Something feels nice about that.

Yet, I don't find the second example objectionable, on its face, although, I'd probably access `User` data via `bidders`, just without the graphy terminology. I think I'm just disappointed in how that terminology intrudes on what's otherwise a very domain-focused schema. But I can see how it's simple to write a "vanilla" relationship field in a way that can be upgraded to a connection, since the vanilla relationship would be an array but the connection is an object. I suppose it could be done with a union type and an argument to choose the collection mode, but then you'd have to use inline fragments to query it. Ew.

Reply:
I do agree and am not particularly fond of the words ‘edges’ and ‘node’, but in the end I’ve come to appreciate that you can write more generic code against standardised terminology more.

If there were a way to be able to remain compatible without initially creating a connection, but you can’t really because with a union you need to add an inline fragment with type selection, as you say.